### PR TITLE
univariate dual vectors for arbitrary sizes

### PR DESCRIPTION
**Pull request by marcocuturi**
this PR implements an extension of the univariate solver that outputs dual vectors, regardless of input sizes / weights.

the implementation uses a lineax sparse solver, instead of going through a manually coded Gauss-seidel loop as in e.g. Alg. 1 of https://arxiv.org/pdf/2201.00730.pdf 

the linear system is mostly build by looking at transported mass (returned by the primal solver) and inverting the corresponding linear system.

at the moment, i still observe a bug for the (arguably) least interesting case when comparing two univariate distributions of the same size (in the test, when weight vectors associated to both `x` and `z` are `unif_n` for instance). note this was the case that the https://arxiv.org/abs/2206.07630 paper was dwelling a bit deeper into, but with a quadratic iteration.

note: it might be interesting to actually code the Alg. 1 of https://arxiv.org/pdf/2201.00730.pdf loop with single entry updates and see if it does run faster than this iterative solver.

### CODE REVIEW

File: tests/solvers/linear/univariate_test.py [hunk 1]
```diff
@@ -28,21 +28,26 @@ class TestUnivariate:
 
   @pytest.fixture(autouse=True)
   def initialize(self, rng: jax.Array):
-    self.rng = rng
+    self.rng = jax.random.PRNGKey(4)
```
Comment (michalk8) [inline tests/solvers/linear/univariate_test.py:31]: Why was this changed?

File: tests/solvers/linear/univariate_test.py [hunk 2]
```diff
@@ -163,3 +168,31 @@ def univ_dist(x, y, a, b):
...
+    n = self.n
+    x, a, y, b, z, c = self.x, self.a, self.y, self.b, self.z, self.c
+    unif_n = jnp.ones((n,)) / n
+    for (target, weights_target) in ((y, b), (z, c)):
+      for weights_source in (a, unif_n):
+        geom = pointcloud.PointCloud(x, target, cost_fn=costs.SqEuclidean())
+        prob = linear_problem.LinearProblem(
+            geom=geom, a=weights_source, b=weights_target
+        )
+        out = univariate.UnivariateSolver()(prob, return_dual_vectors=True)
```
Comment (michalk8) [inline tests/solvers/linear/univariate_test.py:185]: Let's move the solver outside the loops and JIT it.

File: tests/solvers/linear/univariate_test.py [hunk 3]
```diff
@@ -163,3 +168,31 @@ def univ_dist(x, y, a, b):
...
+  @pytest.mark.fast()
+  def test_dual_vectors(self):
+    """The OTT solver outputs meaningful dual variables."""
+    _ = pytest.importorskip("lineax")  # only tested using lineax
+    n = self.n
+    x, a, y, b, z, c = self.x, self.a, self.y, self.b, self.z, self.c
+    unif_n = jnp.ones((n,)) / n
+    for (target, weights_target) in ((y, b), (z, c)):
+      for weights_source in (a, unif_n):
+        geom = pointcloud.PointCloud(x, target, cost_fn=costs.SqEuclidean())
```
Comment (michalk8) [inline tests/solvers/linear/univariate_test.py:181]: Since this function seems to run fast, can be parametrize again by the `cost_fn`?

File: tests/solvers/linear/univariate_test.py [hunk 4]
```diff
@@ -163,3 +168,31 @@ def univ_dist(x, y, a, b):
...
     np.testing.assert_allclose(actual, expected, rtol=tol, atol=tol)
+
+  @pytest.mark.fast()
+  def test_dual_vectors(self):
+    """The OTT solver outputs meaningful dual variables."""
+    _ = pytest.importorskip("lineax")  # only tested using lineax
+    n = self.n
+    x, a, y, b, z, c = self.x, self.a, self.y, self.b, self.z, self.c
+    unif_n = jnp.ones((n,)) / n
+    for (target, weights_target) in ((y, b), (z, c)):
```
Comment (michalk8) [inline tests/solvers/linear/univariate_test.py:179]: In general, am not a fan of this sort of tests; one way to solve this is to `pytest.mark.parametrize(("target", "weights"), [("y", "b"), ...])`
and use `target = getattr(self, target)` to get the correct array in the code.

File: tests/solvers/linear/univariate_test.py [hunk 5]
```diff
@@ -163,3 +168,31 @@ def univ_dist(x, y, a, b):
     expected = univ_dist(x, y, a, b + v_b) - univ_dist(x, y, a, b - v_b)
     actual = 2.0 * jnp.vdot(v_b, grad_b)
     np.testing.assert_allclose(actual, expected, rtol=tol, atol=tol)
+
+  @pytest.mark.fast()
+  def test_dual_vectors(self):
+    """The OTT solver outputs meaningful dual variables."""
+    _ = pytest.importorskip("lineax")  # only tested using lineax
+    n = self.n
```
Comment (michalk8) [inline tests/solvers/linear/univariate_test.py:176]: Personal pet peeve: don't think it's that necessary to have `n = self.n` or the `x, a, ... = self.x, self.a`...

File: tests/solvers/linear/univariate_test.py [hunk 6]
```diff
@@ -163,3 +168,31 @@ def univ_dist(x, y, a, b):
...
+    for (target, weights_target) in ((y, b), (z, c)):
+      for weights_source in (a, unif_n):
+        geom = pointcloud.PointCloud(x, target, cost_fn=costs.SqEuclidean())
+        prob = linear_problem.LinearProblem(
+            geom=geom, a=weights_source, b=weights_target
+        )
+        out = univariate.UnivariateSolver()(prob, return_dual_vectors=True)
+
+        f, g = out.dual_a, out.dual_b
+        dual_obj = jnp.sum(f * weights_source[None, :], axis=1)
```
Comment (michalk8) [inline tests/solvers/linear/univariate_test.py:188]: Might be beneficial to add this as a property of the `UnivariateOutput` instead.

**Review by michalk8 — CHANGES_REQUESTED** [review]

File: src/ott/solvers/linear/univariate.py [hunk 7]
```diff
@@ -318,21 +334,52 @@ def quantile_distance(
...
+    # vector of costs masked by mass transfer. only select active constraints
+    support_cost = (diff_q > 0) * successive_costs
+    # sort and select top n+m-1 to grab non-zero in jit-friendly manner
+    idx_cost = jnp.argsort(support_cost)
+    inv_idx = idx_cost[-n - m + 1:]
+    new_cost = successive_costs[inv_idx]
+    # select indices corresponding to active constraints
+    i_orig_x, i_orig_y = orig_i[inv_idx], orig_j[inv_idx]
+
+    def kkt(dual_ab, ridge_kernel=1e-6):
```
[the remark this answers is not in the document]
Comment (marcocuturi) [reply src/ott/solvers/linear/univariate.py:366]: local computations (indices `i_orig_x`) appear in the function. in addition, because this is just a sparse linear system, i find it more natural to have it right before the call to the solver

File: tests/solvers/linear/univariate_test.py (hunk 6 above)
Comment (marcocuturi) [reply tests/solvers/linear/univariate_test.py:188]: this is just for testing, in principle this `dual_obj` should match exactly `ot_costs` (computed differently), so no need to create a property

File: src/ott/solvers/linear/univariate.py [hunk 8]
```diff
@@ -161,6 +163,7 @@ def __call__(
       self,
       prob: linear_problem.LinearProblem,
       return_transport: bool = True,
+      return_dual_vectors: bool = True,
```
Comment (marcocuturi) [inline src/ott/solvers/linear/univariate.py:166]: i would rather put this to `False` by default, I can only foresee few use cases where this is needed.

File: src/ott/solvers/linear/univariate.py [hunk 9]
```diff
@@ -318,21 +334,52 @@ def quantile_distance(
...
-  return cost, jnp.stack([orig_i, orig_j]), diff_q
+  cost = jnp.sum(successive_costs * diff_q)
+  paired_indices, mass_paired_indices, dual_a, dual_b = [
+      None,
+  ] * 4
+
+  if return_transport or return_dual_vectors:
+    n = x.shape[0]
+
+    i_in_sorted_x_of_quantile = all_values_sorter[i_x_cdf_inv] % n
```
[the remark this answers is not in the document]
Comment (marcocuturi) [reply src/ott/solvers/linear/univariate.py:342]: i would, because of sliced W

File: src/ott/solvers/linear/univariate.py [hunk 10]
```diff
@@ -318,29 +331,120 @@ def quantile_distance(
...
+    orig_j = i_y[i_in_sorted_y_of_quantile][1:]
+    paired_indices, mass_paired_indices = jnp.stack([orig_i, orig_j]), diff_q
+  else:
+    paired_indices, mass_paired_indices = None, None
+
+  return cost, paired_indices, mass_paired_indices
+
+
+# TODO(michalk8): better naming
+def north_west_distance(
```
Comment (marcocuturi) [inline src/ott/solvers/linear/univariate.py:355]: I would suggest `north_west_corner_rule`  (e.g. https://arxiv.org/pdf/1803.00567 3.4.2). There it's in the primal, here we're running it in the dual.

File: src/ott/solvers/linear/univariate.py [hunk 11]
```diff
@@ -318,29 +331,120 @@ def quantile_distance(
...
+  p_final = jnp.maximum(state.a[-1], state.b[-1])
+  mass_paired_indices = state.mass_paired_indices.at[-1].set(p_final)
+
+  return (
+      ot_cost,
+      state.paired_indices,
+      mass_paired_indices,
+      state.dual_a,
+      state.dual_b,
+  )
```
Comment (marcocuturi) [inline src/ott/solvers/linear/univariate.py:440]: thanks, very cool!

**Review by marcocuturi — COMMENTED** [review]
thanks Michal for the implementation of NW

File: src/ott/solvers/linear/univariate.py [hunk 12]
```diff
@@ -240,25 +252,25 @@ def uniform_distance(
...
   Args:
     x: Vector ``[n,]`` of real values.
     y: Vector ``[n,]`` of real values.
     cost_fn: Translation invariant cost function, i.e. ``c(x, y) = h(x - y)``.
-    return_transport: whether to return mapped pairs.
+    return_transport: Whether to return mapped pairs.
 
   Returns:
-    optimal transport cost, a list of ``n+m`` paired indices, and their
+    Optimal transport cost, a list of ``n + m`` paired indices, and their
```
[the remark this answers is not in the document]
Comment (JTT94) [reply src/ott/solvers/linear/univariate.py:265]: In this non regularized setting wont there be at most n+m-1 non-zero entries in the coupling?

File: src/ott/solvers/linear/univariate.py (hunk 11 above)
Comment (JTT94) [reply src/ott/solvers/linear/univariate.py:440]: thanks for tidying up my implementation, I will delete this branch https://github.com/ott-jax/ott/tree/feature/gs_1d_solver

why did the minimum cause a slow down?

File: src/ott/solvers/linear/univariate.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (marcocuturi) [reply src/ott/solvers/linear/univariate.py:265]: my implementation extends the scipy implementation, which leverages search sorted using the union of the two supports. As a result, my implementation outputs pairs of indices + mass, but many (roughly half) of these masses are 0. Hence there are a few "ghost" pair of indices that have no mass. The number of actual non-ghost pairs is upper bounded by `n+m-1`, but it's easier for jitting purposes to carry those `2(n+m)` pairs.

File: src/ott/solvers/linear/univariate.py (hunk 11 above)
Comment (JTT94) [reply src/ott/solvers/linear/univariate.py:440]: Out of interest, why is this appearing as green and not i my prev commit on the branch? Is it squashed or copied and pasted?

**Review by michalk8 — APPROVED** [review]

File: src/ott/solvers/linear/univariate.py [hunk 13]
```diff
@@ -11,336 +11,311 @@
...
-from ott.geometry import costs, pointcloud
 from ott.math import utils as mu
 from ott.problems.linear import linear_problem
 
 __all__ = [
-    "UnivariateOutput", "UnivariateSolver", "uniform_distance",
-    "quantile_distance"
+    "UnivariateOutput",
+    "uniform_distance",
+    "quantile_distance",
```
Comment (marcocuturi) [inline src/ott/solvers/linear/univariate.py:27]: not comfortable with this name of `quantile_distance`, this was used as the name for an inner routine, originating in the original `scipy` implementation, which was only written for the 1-Wasserstein distance. But this has been adapted to fit any `TICost`, therefore not necessarily a distance.

Also, in principle, `north_west_distance` should output the same thing (except for dual vectors) as `quantile_distance`, so this is a bit confusing. Also, shouldn't be called a `distance`, these are fundamentally solvers that return optimal cost / OT primal variable  and possibly (for NW solver) the OT dual variable, not distances

File: src/ott/solvers/linear/univariate.py [hunk 14]
```diff
@@ -11,336 +11,311 @@
...
-  y_pdf = jnp.concatenate([jnp.zeros_like(a), b[i_y]])[all_values_sorter]
 
-  x_cdf = jnp.cumsum(x_pdf)
-  y_cdf = jnp.cumsum(y_pdf)
+def north_west_distance(prob: linear_problem.LinearProblem) -> UnivariateOutput:
+  r"""Distance between two distributions using the north-west corner rule.
 
-  x_y_cdfs = jnp.concatenate([x_cdf, y_cdf])
-  quantile_levels, _ = mu.sort_and_argsort(x_y_cdfs, argsort=False)
+  This function uses the dual form as described in :cite:`sejourne:22`, alg. 3.
```
Comment (marcocuturi) [inline src/ott/solvers/linear/univariate.py:218]: I don't think `dual form` means much here. Maybe reference NW solution in 3.4.2 computational OT book + Séjourné to also store dual variables?

File: src/ott/solvers/linear/univariate.py [hunk 15]
```diff
@@ -11,336 +11,311 @@
...
+      mass_paired_indices=mpi,
   )
 
-  x_pdf = jnp.concatenate([a[i_x], jnp.zeros_like(b)])[all_values_sorter]
-  y_pdf = jnp.concatenate([jnp.zeros_like(a), b[i_y]])[all_values_sorter]
 
-  x_cdf = jnp.cumsum(x_pdf)
-  y_cdf = jnp.cumsum(y_pdf)
+def north_west_distance(prob: linear_problem.LinearProblem) -> UnivariateOutput:
+  r"""Distance between two distributions using the north-west corner rule.
```
Comment (marcocuturi) [inline src/ott/solvers/linear/univariate.py:216]: not a distance :)

File: src/ott/solvers/linear/univariate.py [hunk 16]
```diff
@@ -11,336 +11,311 @@
...
+      :class:`translation-invariant cost <ott.geometry.costs.TICost>`.
+      The ``[n,]`` and ``[m,]`` sized probability weights are stored
+      in attributes :attr:`~ott.problems.linear.linear_problem.LinearProblem.a`
+      and :attr:`~ott.problems.linear.linear_problem.LinearProblem.b`.
 
-  diff_q = jnp.diff(quantile_levels)
-  cost = jnp.sum(
-      jax.vmap(cost_fn.h)(y_cdf_inv[1:, None] - x_cdf_inv[1:, None]) * diff_q
+  Returns:
+    The univariate output. Note that the mass can be null in some entries,
```
Comment (marcocuturi) [inline src/ott/solvers/linear/univariate.py:229]: same comment, I think a reference to `mass_paired_indices` is needed

File: src/ott/solvers/linear/univariate.py (hunk 13 above)
Comment (marcocuturi) [reply src/ott/solvers/linear/univariate.py:27]: I also think, finally that a discussion should be introduced on which implementation to use. I expect the current `quantile_distance` to be faster than `north_west_distance`, hence I would only run the latter if `dual_variables` output is needed, which I don't expect to happen often.

**Review by marcocuturi — COMMENTED** [review]
thanks a lot Michal!! a few questions here and there, about naming but maybe this hides some API choices :)

File: src/ott/solvers/linear/univariate.py (hunk 13 above)
Comment (michalk8) [reply src/ott/solvers/linear/univariate.py:27]: Done, thanks for the comment!
